doc/debugging.rst

```diff
@@ -14,33 +14,108 @@ This allows breakpoints and the execution flow to be viewed and controlled right
+
+.. _console commands: https://pythonconquerstheuniverse.wordpress.com/2009/09/10/debugging-in-python/
+
+.. image:: images/debugging/debugging-arrow.png
+   :alt: Spyder Editor showing the debugging panel
+
+
+Spyder's debugger offers syntax highlighting, code completion and command history, which work exactly like they do in the normal interactive interpreter.
+
+Use the up and down arrows to recall previous commands, and press :kbd:`Tab` to trigger autocomplete suggestions.
+
+.. image:: images/debugging/debugging-autocompletion.gif
+   :alt: A Spyder IPython console window, showing autocompletion when debugging
+
+
+Furthermore, IPython's `magic functions`_ are available in debugging mode.
+You can, for example, run ``%ls`` to list the contents of your current working directory or ``%timeit`` to check how fast a given snippet of code is.
+
+.. _magic functions: https://ipython.readthedocs.io/en/stable/interactive/magics.html
+
+.. image:: images/debugging/debugging-timeit.png
+   :alt: Ipython console in debug mode showing timeit magic
+
+Finally, you can enter and execute multiline statements in Spyder's debugger just like with the regular IPython prompt, to easily run complex code.
+
+.. image:: images/debugging/debugging-multiline.gif
+   :alt: Debugger showing multiline debugger
+
+
+
+===========
+Breakpoints
+===========
+
+Spyder's debugger is integrated with the :guilabel:`Breakpoints` pane, which lists the file, line, and condition (if any) of every breakpoint defined.
+To open it, select :menuselection:`Debug --> List breakpoints`, or press :kbd:`Ctrl-Shift-B` (:kbd:`Cmd-Shift-B` on macOS).
 
 .. image:: images/debugging/breakpoints-standard.png
    :alt: Spyder's Breakpoints panel, with a number of examples showing file, line number and an optional condition
 
-* A :guilabel:`Breakpoints` pane, listing the file, line, and condition (if any) of every breakpoint defined (:menuselection:`&Debug --> List breakpoints`, or :kbd:`Ctrl-Shift-B` by default).
+There are several different ways to set and clear breakpoints:
 
-* Full GUI control over debugger execution from the :guilabel:`Debug` menu, :guilabel:`Debug toolbar` and via configurable keyboard shortcuts, along with the standard ``ipdb`` console commands.
+  * With the :guilabel:`Set/clear breakpoint` option in the Debug menu.
+  * Through pressing the configurable keyboard shortcut (:kbd:`F12` for normal, or :kbd:`Shift-F12` for conditional breakpoints by default).
+  * By clicking to the left of the line number in an open file in the Editor (adding :kbd:`Shift` for a conditional breakpoint).
+  * With the ``breakpoint()`` builtin function in your code.
+  * Interactively, using the ``b`` command in a debugging session.
 
-.. image:: images/debugging/debugging-condbreakpoint.png
-   :alt: Inset of Spyder's Editor, with a breakpoint set and the condition dialog open
+.. image:: images/debugging/debugging-breakpoints.gif
```

Comment:
* [ ] It would be clearer, more useful, and more typical, to show a conditional breakpoint being set inside one of the for loops you have there, to break execution at a certain loop iteration. So you could do, e.g., `b == 4`; assuming you're calling `millerRabin` with argument e.g. `m >= 6`. 
* [ ] I also think it useful to actually run the debugger with the Variable Explorer open) to show that it is stopped when `b` equals that value; it should only take a few more seconds to run (just press the debug button and open the Variable Explorer, showing the `b` value) and would illustrate the effect of the condition breakpoint.

Reply:
Agreed with both suggestions.

Reply:
This looks good to me now, thanks.